web-api/src/mind.rs

```diff
@@ -231,6 +315,124 @@ where
+    let scores = documents
+        .iter()
+        .map(|document| {
+            if document.is_interesting(user_interests) {
+                2.0
+            } else {
+                0.0
+            }
+        })
+        .collect::<Vec<f32>>();
+    Array1::from(scores)
+}
+
+fn write_array<T, S, D>(writer: impl io::Write, array: &ndarray::ArrayBase<S, D>) -> io::Result<()>
+where
+    T: Clone + npyz::AutoSerialize,
+    S: ndarray::Data<Elem = T>,
+    D: ndarray::Dimension,
+{
+    let shape = array.shape().iter().map(|&x| x as u64).collect::<Vec<_>>();
+    let c_order_items = array.iter();
+
+    let mut writer = npyz::WriteOptions::new()
+        .default_dtype()
+        .shape(&shape)
+        .writer(writer)
+        .begin_nd()?;
+    writer.extend(c_order_items)?;
+    writer.finish()
+}
+
+/// Runs persona based benchmark
+async fn run_persona_based_benchmark() -> Result<(), Error> {
+    let document_provider = DocumentProvider::new("news.tsv")?;
+    let state = State::new(Storage::default()).unwrap();
+    // load documents from document provider to state
+    state
+        .insert(
+            document_provider
+                .documents
+                .values()
+                .cloned()
+                .collect::<Vec<_>>(),
+        )
+        .await
+        .unwrap();
+    let benchmark_config = PersonaBasedConfig::default();
+    // create 3d array of zeros with shape (nranks, iterations, n_documents)
+    let mut results = Array3::<f32>::zeros(Dim([
+        benchmark_config.nranks.len(),
+        benchmark_config.iterations,
+        benchmark_config.n_documents,
+    ]));
+
+    let mut rng = thread_rng();
+
+    let users_interests = Users::new("user_categories.json")?;
+    for (idx, (user_id, interests)) in users_interests.0.iter().enumerate() {
+        let interesting_documents = document_provider.get_all_interest(interests);
+        let ids_of_documents_to_prepare = interesting_documents
+            .choose_multiple(&mut rng, benchmark_config.amount_of_doc_used_to_prepare)
+            .map(|doc| doc.id.clone())
+            .collect::<Vec<_>>();
+        // prepare reranker by interacting with documents to prepare
+        state
+            .interact(user_id, &ids_of_documents_to_prepare)
+            .await
+            .unwrap();
+        // perform iterations
+        for iter in 0..benchmark_config.iterations {
+            // get personalised documents from state
+            let personalised_documents = state
+                .personalize(
+                    user_id,
+                    PersonalizeBy::KnnSearch(benchmark_config.n_documents),
+                )
+                .await
+                .unwrap();
+            // get documents based on the personalised documents ids
+            let documents = personalised_documents
+                .iter()
+                .map(|id| document_provider.get(id).unwrap())
+                .collect::<Vec<_>>();
```

Comment:
:art: the `Itertools` trait provides a nice shorthand for this (it doesn't allow for type hints for the elements though, if that's an issue)
```suggestion
                .collect_vec();
```